src/buildercore/cfngen.py

```diff
@@ -237,6 +238,18 @@ def build_subdomain(x):
     else:
         context['cloudfront'] = False
 
+def build_context_fastly(context, parameterize):
+    def build_subdomain(x):
+        return complete_domain(parameterize(x), context['domain'])
+    if 'fastly' in context['project']['aws']:
+        context['fastly'] = {
+            'subdomains': [build_subdomain(x) for x in context['project']['aws']['fastly']['subdomains']],
+            # future use
+            'subdomains-without-dns': [],
```

Comment:
probably lots of stuff will appear here, with Fastly terminology. We don't do a translation from abstract terms, e.g. everything under `cloudfront` is pretty much taken from CloudFront/CloudFormation documentation. I expect `fastly` to be similar, otherwise the user trying to configure it has to learn yet another terminology on top of the CloudFront/Fastly one. Portability of the same configuration across cloud providers or vendors is out of scope, don't think even Terraform does that as AWS stuff is specified through the [aws provider](https://www.terraform.io/docs/providers/aws/).

`subdomains` is a rare example of an abstract key.

Reply:
> Portability of the same configuration across cloud providers or vendors is out of scope,

:+1:

In the beginning with cloudformation I kept project configuration abstract and translated options into cfn features collected together in cfngen.py which was then used to render a simple template with python formatting strings. then jinja and then finally troposphere was introduced to generate the template. It did mean that while the project property you were changing was easy to do, tracing it from project config into template meant traversing through a lot of code. Gathering this config under specific headings like `fastly` or `rds` rather than `cdn` or `remote-database` means we can skip the code traversal (probably) and match the options to the docs. The next step might be a simple pass-through of options in project config into the final template.

Reply:
One important thing that happens here is interpolation of strings e.g. `pattern-library-{instance}` bucket name being interpolated to `pattern-library-ci`